runtime/near-vm-runner/src/near_vm_runner.rs

```diff
@@ -242,11 +242,27 @@ impl NearVM {
         let compiler = Singlepass::new();
         // We only support universal engine at the moment.
         assert_eq!(VM_CONFIG.engine, NearVmEngine::Universal);
+
+        // FIXME: should have as many code memories as there are possible parallel invocations of
+        // the runtime… How do we determine that?
+        static CODE_MEMORY_POOL_CELL: OnceLock<LimitedMemoryPool> = OnceLock::new();
+        let code_memory_pool = CODE_MEMORY_POOL_CELL
+            .get_or_init(|| {
+                LimitedMemoryPool::new(128, 256 * 1024).unwrap_or_else(|e| {
```

Comment:
why 256 kiB?
(and 128 memories seems a bit on the large side but I guess that's what the comment above is about)

Reply:
256kiB is just my best guess for the size of code memory that will not require re-allocating for most of the contracts. Ideally this needs somebody (me!) to grab all contracts and see how large they are loaded up. We could also go for 4MiB (WASM contract size limit) or some multiple of that.

128 is a best guess for the number of parallel invocations of a runtime we might have in a 100 shard world. Anything above will lead to memory maps being allocated on the fly, which is kind of the worst case. This is also why I was asking around for what we think is the maximum parallelism we might see. (Another observation is that there aren't too many system configurations with ≥128 cores…)

Reply:
Yeah makes sense. Maybe a comment regarding 256kB being a best guess would be useful.

Reply:
4MiB, unfortunately, isn’t the maximum a compiled loadable binary might occupy. That's the limit of the size of WebAssembly code. I would imagine that there are short sequences of WebAssembly code that expand to a non-trivial number of x86_64 instructions that would then occupy much more than the initial WebAssembly instruction. In that sense 256KiB is as good of a choice as 4MiB is.

If we wanted to come up with a precise upper bound, we would need to do some non-trivial analysis on our singlepass backend to figure out what the maximum expansion factor is for a given target.

Reply:
I guess my tldr is, I’d feel more comfortable having an upper bound (even if it means eg. 64MB and reducing the number of memories to 8; and poking them while creating them so they don’t get overallocated).

The reason is, it’s very easy for literally any attacker to force all memories to grow to the upper bound. On the contrary, forcing forks with multiple shards to have us process multiple contracts concurrently is significantly harder, so the number of code memories is much less important.

And maybe in a second step we should introduce a protocol update and make the code memories non-growable, rejecting the contract if it compiles to more than the preconfigured size, as a second layer of defense.

Does that make sense?

Reply:
I largely agree that having an upper bound would be great. I don’t think right now we're in a position to set one such up, though. But your observation about one being able to eventually grow the entire pool of memories to be much larger is true, and I think it is indeed appropriate to limit the number of pre-allocated memories to a much lower number.

> And maybe in a second step we should introduce a protocol update and make the code memories non-growable, rejecting the contract if it compiles to more than the preconfigured size

In the current scheme of things, that would mean we wouldn’t be able to modify codegen without making protocol version bumps. For this idea to materialize we'd need limited replayability plus implicit protocol version bumps on every release.